### PR TITLE
Access Delivery Session Appointments

### PR DESCRIPTION
**Pull request by andrewburrowsmoj**
This PR is to facilitate the ability to show all appointments across all sessions on referral progress screen.

Currently we only show the latest appointment for a session and the existing endpoints only consider the latest; furthermore the endpoints tie a session to an action plan rather than referral.

Also need the ability to get an appointment based on appointment id rather than session id since session id only makes sense if you want the latest appointment.

commit 811aaa179025cbb368e77428853554adb50573e9

Add new endpoint /referral/{referralId}/delivery-session-appointment/{appointmentId}

Allows for access to an appointment based on referral and delivery session appointment id rather than action plan.
Returns a DeliverySessionAppointment object which is different from a DeliverySession in that the id is the appointment.

The DeliverySession object contains the latest appointment for the session which can be a bit misleading as the id is the session id rather than the appointment id.

commit f109c44e41798e6fb4d645f995ba068e7df987c3
New endpoint GET /referral/{referralId}/delivery-session-appointments

This returns all appointments for all sessions of a referral; this includes appointments that have been rescheduled for the same session.
Existing endpoints connected with delivery session would only return the latest appointment for a session.

commit de13714de363169a6ec7e535e51407387f7b2dc9 
Added two new endpoints:

POST /referral/{referralId}/delivery-session-appointment - allows to scheduling a new appointment for a session provided that the existing appointments (if any) have feedback submitted
PATCH /referral/{referralId}/delivery-session-appointment/{appointmentId} - allows to reschedule an existing appointment provided that no feedback has been submitted

Deprecated the existing endpoint PATCH /action-plan/{id}/appointment/{sessionNumber}. This endpoint performs both scheduling and rescheduling based on action plan id and session number.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt [hunk 1]
```diff
@@ -118,4 +126,22 @@ class DeliverySessionController(
     val user = userMapper.fromToken(authentication)
     return DeliverySessionDTO.from(deliverySessionService.submitSessionFeedback(actionPlanId, sessionNumber, user))
   }
+
+  @GetMapping("/referral/{referralId}/delivery-session-appointment/{appointmentId}")
```
Comment (andrewburrowsmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt:136]: make this resource noun plural

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt [hunk 2]
```diff
@@ -161,4 +164,56 @@ class DeliverySessionController(
...
+      user,
+      newDeliverySessionAppointmentRequest.appointmentDeliveryType,
+      newDeliverySessionAppointmentRequest.sessionType,
+      newDeliverySessionAppointmentRequest.appointmentDeliveryAddress,
+      newDeliverySessionAppointmentRequest.npsOfficeCode,
+    )
+    return DeliverySessionAppointmentDTO.from(deliverySession.sessionNumber, deliverySession.currentAppointment!!)
+  }
+
+  @PatchMapping("/referral/{referralId}/delivery-session-appointment/{appointmentId}")
```
Comment (andrewburrowsmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt:193]: this should be PUT

Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt:193]: make plural

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt [hunk 3]
```diff
@@ -161,4 +164,56 @@ class DeliverySessionController(
     return deliverySessionService.getSessions(referralId)
       .flatMap { session -> session.appointments.map { appointment -> DeliverySessionAppointmentDTO.from(session.sessionNumber, appointment) } }
   }
+
+  @PostMapping("/referral/{referralId}/delivery-session-appointment")
```
Comment (andrewburrowsmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt:168]: make plural

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/DeliverySessionService.kt [hunk 4]
```diff
@@ -62,6 +62,84 @@ class DeliverySessionService(
...
+    appointmentDeliveryType: AppointmentDeliveryType,
+    appointmentSessionType: AppointmentSessionType,
+    appointmentDeliveryAddress: AddressDTO? = null,
+    npsOfficeCode: String? = null,
+  ): DeliverySession {
+    val session = getDeliverySession(referralId, sessionNumber) ?: throw EntityNotFoundException("Session not found for referral [referralId=$referralId, sessionNumber=$sessionNumber]")
+    val existingAppointment = session.currentAppointment
+
+    existingAppointment?.let {
+      if (it.appointmentTime.isAfter(appointmentTime)) {
```
Comment (andrewburrowsmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/DeliverySessionService.kt:80]: should this be in reschedule instead?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt (hunk 3 above)
Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt]: done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt (hunk 2 above)
Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt]: done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt (hunk 1 above)
Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt]: done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt [hunk 5]
```diff
@@ -118,4 +133,88 @@ class DeliverySessionController(
...
+    @PathVariable(name = "referralId") referralId: UUID,
+    @PathVariable(name = "appointmentId") appointmentId: UUID,
+    authentication: JwtAuthenticationToken,
+  ): DeliverySessionAppointmentDTO {
+    val user = userMapper.fromToken(authentication)
+    val referral = referralService.getSentReferral(referralId) ?: throw ResponseStatusException(
+      HttpStatus.BAD_REQUEST, "sent referral not found [referralId=$referralId]"
+    )
+    referralAccessChecker.forUser(referral, user)
+    val matchingAppointment = deliverySessionService.getSessions(referralId)
```
Comment (mahalrmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt:148]: Is this not model/business related logic and hence reusable as a service method?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt [hunk 6]
```diff
@@ -118,4 +133,88 @@ class DeliverySessionController(
...
+    val referral = referralService.getSentReferral(referralId) ?: throw ResponseStatusException(
+      HttpStatus.BAD_REQUEST, "sent referral not found [referralId=$referralId]"
+    )
+    referralAccessChecker.forUser(referral, user)
+    return deliverySessionService.getSessions(referralId)
+      .flatMap { session -> session.appointments.map { appointment -> DeliverySessionAppointmentDTO.from(session.sessionNumber, appointment) } }
+  }
+
+  @PostMapping("/referral/{referralId}/delivery-session-appointments")
+  fun scheduleNewDeliverySessionAppointment(
```
Comment (mahalrmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt:170]: Would the word "initial" be clearer as "new" appointments are event created when rescheduling

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/DeliverySessionAppointmentDTO.kt [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  val npsOfficeCode: String?,
+  val appointmentDeliveryAddress: AddressDTO?,
+  val sessionFeedback: SessionFeedbackDTO,
+) {
+  companion object {
+    fun from(sessionNumber: Int, appointment: Appointment): DeliverySessionAppointmentDTO {
+      val appointmentDelivery = appointment.appointmentDelivery
+      val address = when (appointmentDelivery?.appointmentDeliveryType) {
+        AppointmentDeliveryType.IN_PERSON_MEETING_OTHER -> {
+          if (appointmentDelivery?.appointmentDeliveryAddress !== null) {
```
Comment (mahalrmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/DeliverySessionAppointmentDTO.kt:25]: should we be using ?.let syntax?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/DeliverySessionService.kt [hunk 8]
```diff
@@ -61,6 +62,84 @@ class DeliverySessionService(
     return deliverySessionRepository.save(session)
   }
 
+  fun scheduleNewDeliverySessionAppointment(
```
Comment (mahalrmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/DeliverySessionService.kt:65]: Again "initial" over "new"?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/DeliverySessionService.kt [hunk 9]
```diff
@@ -61,6 +62,84 @@ class DeliverySessionService(
...
+    appointmentDeliveryType: AppointmentDeliveryType,
+    appointmentSessionType: AppointmentSessionType,
+    appointmentDeliveryAddress: AddressDTO? = null,
+    npsOfficeCode: String? = null,
+  ): DeliverySession {
+    val session = getDeliverySession(referralId, sessionNumber) ?: throw EntityNotFoundException("Session not found for referral [referralId=$referralId, sessionNumber=$sessionNumber]")
+    val existingAppointment = session.currentAppointment
+
+    existingAppointment?.let {
+      if (it.appointmentTime.isAfter(appointmentTime)) {
```
Comment (mahalrmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/DeliverySessionService.kt:80]: Why is this necessary? Isn't it covered by the check that the latest appointment has feedback

**Review by mahalrmoj — APPROVED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt (hunk 6 above)
Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt:170]: This is either the first (delivery session) appointment or any additional appointment.
If additional then it checks that the existing appointment is before this new one and that the feedback was submitted.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/DeliverySessionService.kt (hunk 9 above)
Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/DeliverySessionService.kt:80]: No because the time given for this new appointment could be before the existing one even though feedback was submitted.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt (hunk 5 above)
Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt:148]: In this case we are transforming the `Session`s' list of appointments into a list of appointment and sessionNumber pairs.

It's not possible to delegate this functionality into the service method since the sessionNumber is not retrievable from an `Appointment` object. We can't know which `Session` the retrieved `Appointment` belongs to without using some type of wrapper object; an `Appointment` has no concept of whether it's an initial assessment or a delivery session appointment.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/DeliverySessionAppointmentDTO.kt (hunk 7 above)
Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/DeliverySessionAppointmentDTO.kt]: Thanks. I've used this syntax now.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt [hunk 10]
```diff
@@ -118,4 +133,88 @@ class DeliverySessionController(
...
+    authentication: JwtAuthenticationToken,
+  ): DeliverySessionAppointmentDTO {
+    val user = userMapper.fromToken(authentication)
+    val referral = referralService.getSentReferral(referralId) ?: throw ResponseStatusException(
+      HttpStatus.BAD_REQUEST, "sent referral not found [referralId=$referralId]"
+    )
+    referralAccessChecker.forUser(referral, user)
+    val matchingAppointment = deliverySessionService.getSessions(referralId)
+      .flatMap { session -> session.appointments.map { appointment -> Pair(session.sessionNumber, appointment) } }
+      .firstOrNull { appointment -> appointment.second.id == appointmentId }
```
Comment (sldblog) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt:150]: @andrewburrowsmoj @mahalrmoj In non-Boot environments, I've seen this pattern dealt by using custom queries in the service, which feels like a more efficient approach.

I don't quite understand the need for this endpoint though -- `/referral/{id}/sessions/{sessionNumber}` seems ok and doesn't use action plans. Is this endpoint necessary?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt [hunk 11]
```diff
@@ -118,4 +133,88 @@ class DeliverySessionController(
...
+    @PathVariable(name = "referralId") referralId: UUID,
+    authentication: JwtAuthenticationToken,
+  ): List<DeliverySessionAppointmentDTO> {
+    val user = userMapper.fromToken(authentication)
+    val referral = referralService.getSentReferral(referralId) ?: throw ResponseStatusException(
+      HttpStatus.BAD_REQUEST, "sent referral not found [referralId=$referralId]"
+    )
+    referralAccessChecker.forUser(referral, user)
+    return deliverySessionService.getSessions(referralId)
+      .flatMap { session -> session.appointments.map { appointment -> DeliverySessionAppointmentDTO.from(session.sessionNumber, appointment) } }
```
Comment (sldblog) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/DeliverySessionController.kt:166]: `deliverySessionService.getSessionAppointments(referralId)` sounds like something that wouldn't force `deliverySessionService` users to transform on their own -- is this a smell that our entity/data model isn't fit for purpose?
